pkg/resource/deploy/source_eval.go

```diff
@@ -522,6 +534,24 @@ func (d *defaultProviders) getDefaultProviderRef(req providers.ProviderRequest)
 	return res.ref, res.err
 }
 
+func (d *defaultProviders) setProviderAsDefault(ref providers.Reference, pkg, version, pluginDownloadURL string) error {
+	if version != "" {
+		version = "-" + version
+	}
+	if pluginDownloadURL != "" {
+		pluginDownloadURL = "-" + pluginDownloadURL
+	}
+	key := pkg + version + pluginDownloadURL
+	// TODO: this is a bit awkward with versioning.  E.g. should we also set `d.providers[pkg] = ref` here?
```

Comment:
I think this is a critical aspect. I observe that the default provider may or may not have a version component, depending on which SDK is in use. I would argue that there should be one default provider per package, not per package+version.

Reply:
I think it needs to be per package+version. Since users can pass in the version on resource registration, the following would not work otherwise:

```
		_, err = random.NewRandomPet(ctx, "example", nil, pulumi.Version("4.2.0"))
		if err != nil {
			return err
		}
		_, err = random.NewRandomPet(ctx, "example2", nil, pulumi.Version("4.3.0"))
		if err != nil {
			return err
		}
```

Reply:
> I would argue that there should be one default provider per package

Thomas is correct, this definitely needs to be per pkg+version. The fixes in https://github.com/pulumi/pulumi/pull/16109 remove the "unversioned" default provider, and that might be ok as well, need to check it doesn't break anything odd backcompat wise.